pennylane/devices/qubit/adjoint_jacobian.py

```diff
@@ -218,6 +214,108 @@ def adjoint_jvp(tape: QuantumTape, tangents: Tuple[Number], state=None):
     return tuple(np.array(t) for t in tangents_out)
 
 
+def _get_vjp_bras(tape, cotangents, ket):
+    """Helper function for getting the bras for adjoint vjp, the batch size of the
+    cotangents, as well as a list of indices for which the cotangents are zero.
+
+    Args:
+        tape (QuantumTape): circuit that the function takes the gradient of.
+        tangents (Tuple[Number]): gradient vector for input parameters.
+        ket (TensorLike): the final state of the circuit.
+
+    Returns:
+        Tuple[TensorLike, int, List]: The return contains the following:
+            * Final bra for batch size ``None``, else array of bras
+            * Batch size. None if cotangents are not batched
+            * List containing batch indices that are zero. Empty for unbatched
+              cotangents
+    """
+
+    if isinstance(tape.measurements[0], qml.measurements.StateMP):
+        batched_cotangents = qml.math.ndim(cotangents) == 2
+        batch_size = qml.math.shape(cotangents)[0] if batched_cotangents else None
+        bras = np.conj(cotangents.reshape(-1, *ket.shape))
```

Comment:
do we know that `cotangents` is a numpy value at this point? you're using a mix of `np` and `qml.math`

Reply:
Yes. it is strictly numpy and floats here.

Reply:
nitpick but i'd prefer that we use `np` instead of `qml.math` to make it extra obvious, and it's probably more performant (though marginal at best)

Reply:
With the new changes, since we're using the same pipeline for device VJP with and without jitting, the cotangents should be jax arrays

Reply:
Also makes it more clear that this is strictly numpy here.